cpp/src/io/parquet/predicate_pushdown.cpp

```diff
@@ -444,55 +309,180 @@ std::optional<std::vector<std::vector<size_type>>> aggregate_reader_metadata::fi
     columns.push_back(std::move(min_col));
     columns.push_back(std::move(max_col));
   }
-  auto stats_table = cudf::table(std::move(columns));
 
-  // Converts AST to StatsAST with reference to min, max columns in above `stats_table`.
-  stats_expression_converter const stats_expr{filter.get(),
-                                              static_cast<size_type>(output_dtypes.size())};
-  auto stats_ast     = stats_expr.get_stats_expr();
-  auto predicate_col = cudf::detail::compute_column(stats_table, stats_ast.get(), stream, mr);
-  auto predicate     = predicate_col->view();
-  CUDF_EXPECTS(predicate.type().id() == cudf::type_id::BOOL8,
-               "Filter expression must return a boolean column");
+  // Apply bloom filter membership to row groups for each equality predicate column, literal pair
+  // and get corresponding boolean columns.
+  auto [bloom_filter_membership_cols, equality_literals] =
+    apply_bloom_filters(sources,
+                        input_row_group_indices,
+                        output_dtypes,
+                        output_column_schemas,
+                        filter,
+                        total_row_groups,
+                        stream);
+
+  // Check if we have any bloom filter membership columns
+  auto const has_bloom_filters = not bloom_filter_membership_cols.empty();
+
+  // Append bloom filter membership columns to stats columns to get combined columns if needed
+  if (has_bloom_filters) {
+    columns.insert(columns.end(),
+                   std::make_move_iterator(bloom_filter_membership_cols.begin()),
+                   std::make_move_iterator(bloom_filter_membership_cols.end()));
+  }
 
-  auto const host_bitmask = [&] {
-    auto const num_bitmasks = num_bitmask_words(predicate.size());
-    if (predicate.nullable()) {
-      return cudf::detail::make_host_vector_sync(
-        device_span<bitmask_type const>(predicate.null_mask(), num_bitmasks), stream);
-    } else {
-      auto bitmask = cudf::detail::make_host_vector<bitmask_type>(num_bitmasks, stream);
-      std::fill(bitmask.begin(), bitmask.end(), ~bitmask_type{0});
-      return bitmask;
-    }
-  }();
+  auto combined_table = cudf::table(std::move(columns));
 
-  auto validity_it = cudf::detail::make_counting_transform_iterator(
-    0, [bitmask = host_bitmask.data()](auto bit_index) { return bit_is_set(bitmask, bit_index); });
+  // Convert AST to a CombinedAST (StatsAST and BloomfilterAST) expression
+  combined_expression_converter combined_expr{
```

Comment:
`combined_expression_converter` combines both StatsAST and BloomfilterAST converters